File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-search/src/state/searchState.ts
import type { SearchPageResult } from "../types/searchModels.js";

export interface SearchState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: SearchPageResult<unknown>;
  readonly error?: string;
}

export function initialSearchState(): SearchState {
  return { status: "idle" };
}

export function reduceSearchState(
  state: SearchState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: SearchPageResult<unknown> }
    | { type: "failed"; error: string },
): SearchState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}

File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_search/lib/src/state/search_state.dart
import '../models/search_models.dart';

enum SearchStatus { idle, loading, ready, error }

class SearchState {
  const SearchState({
    this.status = SearchStatus.idle,
    this.lastResult,
    this.error,
  });

  final SearchStatus status;
  final SearchPageResult<Object?>? lastResult;
  final String? error;

  SearchState loading() => const SearchState(status: SearchStatus.loading);

  SearchState loaded(SearchPageResult<Object?> result) =>
      SearchState(status: SearchStatus.ready, lastResult: result);

  SearchState failed(String message) =>
      SearchState(status: SearchStatus.error, error: message);
}
